lib/models/addon.js

```diff
@@ -29,13 +31,59 @@ var ensurePosixPath = require('ensure-posix-path');
 var defaultsDeep = require('ember-cli-lodash-subset').defaultsDeep;
 var Babel = require('broccoli-babel-transpiler');
 var findAddonByName = require('../utilities/find-addon-by-name');
+var experiments = require('../experiments');
+var heimdall = require('heimdalljs');
+var calculateCacheKeyForTree = require('../utilities/addon-calculate-cache-key-for-tree');
+
+heimdall.registerMonitor('addon-tree-cache', function AddonTreeCacheSchema() {
+  this.hits = 0;
+  this.misses = 0;
+  this.adds = 0;
+});
 
 var DEFAULT_BABEL_CONFIG = {
   modules: 'amdStrict',
   moduleIds: true,
   resolveModuleSource: require('amd-name-resolver').moduleResolve
 };
 
+var DEFAULT_TREE_FOR_METHODS = {
```

Comment:
In this case, the default `Addon.prototype.cacheKeyForTree` implementation would detect that there is no default method on the `addonProto` object and `return null;` (causing it to be uncached).  This is *precisely* the scenario where an addon (ember-engines in this case) would want to define a custom `cacheKeyForTree` and ensure that it incorporates whatever things would cause the cache key to be invalidated properly.